### PR TITLE
improve generation of diagnostics, fix linter tests

### PR DESCRIPTION
**Pull request by phryneas**
(feat) improve generation of diagnostics
    
>    when the options `checkSyntacticErrors: false` and `useTypescriptIncrementalApi: true` both were active, no semantic errors were emitted as soon as the first syntactic error was encountered
>    this patches the `typescript` import to override that behavior
>    see discussion in #257

(fix) linter tests
    
>    actually use the correct `useTypescriptIncrementalApi` value for some of the linter tests that were using the default value before

(fix) linter tests that were doing nothing
    
>    fix a bug where, as `tslint` was set to true and the files being checked had no parent folder with a tslint.json, tslint never found any warnings
>    also fixes a failing test because a syntactic error was counted as linter warning

see discussion in #257

this is based off #253, so please merge that before you merge this :) (should also make the diff more readable)

### CODE REVIEW

**Comment by phryneas** [comment]
(just force-pushed everything because I had the wording of the commit messages wrong :grin: )

**Comment by piotr-oles** [comment]
I'm afraid you are missing colon next to the `type` section (should be `fix(tests): fix linter tests that were doing nothing`) :) The history of commits is clear so I wouldn't like to squash them :)

File: src/patchTypescript.ts
Comment (piotr-oles) [inline src/patchTypescript.ts]: Should we copy entire typescript object? I think it would be more elegant to copy just methods that we will use in override :)
```
const { 
  createEmitAndSemanticDiagnosticsBuilderProgram: originalCreateEmitAndSemanticDiagnosticsBuilderProgram
} = typescript;
```

File: src/patchTypescript.ts
Comment (piotr-oles) [inline src/patchTypescript.ts]: I'm not sure if `this` inside `createEmitAndSemanticDiagnosticBuilderProgram` will be correct (as we have typescript and origTypescript). I would feel more comfortable with explicit `this` binding:
```
const program = originalCreateEmitAndSemanticDiagnosticsBuilderProgram.apply(typescript, args);
```

File: src/service.ts
Comment (piotr-oles) [inline src/service.ts]: Isn't this more like `patchConfig`?

File: test/integration/mocks/ApiIncrementalCheckerWithRpc.js
Comment (piotr-oles) [inline test/integration/mocks/ApiIncrementalCheckerWithRpc.js]: Please change name to the `originalImport` :)

**Review by piotr-oles — CHANGES_REQUESTED** [review]
Great work :)

File: src/patchTypescript.ts
Comment (phryneas) [reply src/patchTypescript.ts]: For now yes, but #231 uses this pattern for about 20 methods (the file there is currently named wrapTypescript, I'll rename that there to be the same file as this one), so we already know that going forward, this will make a lot of method copies and take up like 20-40 lines (depending on max char width) that way.
The spread on the other hand is just an object full of references to the original implementations - what isn't overridden doesn't really take extra memory, so I'd suggest to go with this from the beginning.

File: src/patchTypescript.ts
Comment (phryneas) [reply src/patchTypescript.ts]: From what I've seen, the typescript code doesn't really make use of the `this` keyword at all, they just have that one reference to `ts` and a lot of nested closures (which is why we can't just wrap it, but have to override some methods directly on the original object).

But yeah, apart from that, this is more future-proof if they should change that at any time in the future. I'll change it to `apply` :+1:

File: src/service.ts
Comment (phryneas) [reply src/service.ts]: Argh, yes. Remnant of an older #231 approach that still lurks my mind :laughing:  Gonna change it :)

**Comment by phryneas** [comment]
> I'm afraid you are missing colon next to the `type` section (should be `fix(tests): fix linter tests that were doing nothing`) :) The history of commits is clear so I wouldn't like to squash them :)

Whoops! Will do a rebase & rename :)

File: src/patchTypescript.ts
Comment (phryneas) [reply src/patchTypescript.ts]: PS: actually, this exact method would need to be overwritten/wrapped a second time for 231, so I guess what I said here doesn't apply there. Gonna move it into a single variable ;)

Comment (phryneas) [reply src/patchTypescript.ts]: done as requested ;)

**Review by piotr-oles — APPROVED** [review]

**Comment by phryneas** [comment]
Okay, I'll rebase this on the current `beta` and then we're good to go.

**Comment by phryneas** [comment]
@piotr-oles  @johnnyreilly  Oh, seeing that I can now merge PRs myself: what is my mode of operation here?

Wait for one of you two to approve the changes, then merge into `beta`? Wait for both of you to approve?

**Comment by piotr-oles** [comment]
For me one approve is enough for fixes and small improvements :) For breaking changes and new features I think it will be better for all of us to get review from two different developers :)

**Comment by johnnyreilly** [comment]
> For me one approve is enough for fixes and small improvements :) For breaking changes and new features I think it will be better for all of us to get review from two different developers :)

That's a sensible general approach I think. 👍  The flip side of that is that it's important that reviews happen in a timely fashion. So if someone has asked me for a review and I haven't come back to them in a reasonable period of time, they should feel free to keep shipping.😁🛳️

BTW feel free to merge this when you've added the comment I requested @phryneas!

File: src/patchTypescript.ts
Comment (johnnyreilly) [inline src/patchTypescript.ts]: Could you put a comment in this module that explains what the purpose of this code is and links back to the GitHub issue discussion that drove this please?

I want to be kind to future maintainers 😁

**Review by johnnyreilly — DISMISSED** [review]

File: src/patchTypescript.ts
Comment (phryneas) [reply src/patchTypescript.ts]: that file has now almost as much documentation as code :grin:

**Comment by johnnyreilly** [comment]
This is great work @phryneas!

@piotr-oles what's the process now? We raise a PR to merge from `beta` to `master`?

**Pull request merged**
